File: 367-valid-perfect-square.cs
public class Solution {
    public bool IsPerfectSquare(int num) {
        if (num < 2) {
            return true;
            }

            long left = 2, right = num / 2 + 1, x, guessSquared;
            while (left < right) {
            x = left + (right - left) / 2;
            guessSquared = x * x;
            if (guessSquared == num) {
                return true;
            }
            if (guessSquared > num) {
                right = x;
            } else {
                left = x + 1;
            }
            }
            return false;
    }
}

File: 367-valid-perfect-square.go
package main

func isPerfectSquare(num int) bool {
	left, right := 0, num+1
	for left < right {
		mid := left + (right-left)/2
		square := mid * mid
		if square < num {
			left = mid + 1
		} else if square > num {
			right = mid
		} else {
			return true
		}
	}
	return false
}
